general/hash.hpp

```diff
@@ -591,6 +596,7 @@ void HashTable<T>::Alloc(int id, int p1, int p2)
       item.p2 = p2;
 
       Insert(Hash(p1, p2), id, item);
+      CheckRehash();
```

Comment:
Shouldn't this line rather be inside the `Insert` method?

Reply:
No, Insert is needed elsewhere without the rehash.

Reply:
Can you describe in the documentation why, and also add documentation to `Insert`, `Hash`, and `CheckRehash`?

Reply:
Sorry, no.

Reply:
Sure: Insert simply adds an item into a bin, it shouldn't do anything else. Also it is used in situations where an item is deleted and reinserted elsewhere (see Reparent), so enlarging the hash table is not needed in that case.

Reply:
No, the logic is "do one thing", "don't mix functionality". Also rehash is expensive, it simply doesn't belong in Insert. And it's not the interface, it's a private method.

Reply:
If I understand correctly the code in `CheckRehash`, it only rehashes when needed, but it's difficult to tell with this undocumented `mask`. Also how comes you `CheckRehash` after you `Insert`, this seems counter-intuitive to me, I would expect an algorithm to check if there is space before inserting, not after.

The idea to put `CheckRehash` inside `Insert` is actually to have `Insert` always do one thing: insert. Currently `Insert` is doing two things: inserting when there is space, creating a bug when there isn't enough space (the reason for this PR).
If you want the unsafe capability I think it would be better to consider a design like the following in the future:
```c++
/// @brief Inserts without checking if there is the necessary space.
template<typename T>
inline void HashTable<T>::UnsafeInsert(int idx, int id, T &item)
{
   // add item at the beginning of the linked list
   item.next = table[idx];
   table[idx] = id;
}

template<typename T>
inline void HashTable<T>::Insert(int idx, int id, T &item)
{
   CheckRehash();
   UnsafeInsert(idx, id, item);
}
```

Reply:
I threw a comment in there to clarify the Insert method for future developers.

Reply:
@acfisher @jakubcerveny @tzanio I pushed documentation for the `HashTable` class, tell me if you think it is clear.